File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int lenOfLL(ListNode *head){
        int len = 0;
        ListNode* temp = head;
        while (temp){
            len += 1;
            temp = temp->next;
        }

        return len;
    }
    ListNode* rotateRight(ListNode* head, int k) {
        if (head == nullptr || head->next == nullptr) return head;
        int len = lenOfLL(head);

        k %= len;
        if (k == 0) return head;
        ListNode *slow = head, *fast = head;

        while (k--){
            fast = fast->next;
        }

        while (fast->next != nullptr){
            slow = slow->next;
            fast = fast->next;
        }

        ListNode *temp = slow->next;
        slow->next = nullptr;
        fast = temp;

        while (fast->next != nullptr){
            fast = fast->next;
        }

        fast->next = head;


        return temp;
    }
};

File: 61-rotate-list/rotate-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func lengthOfLL(head *ListNode) int {
    length := 0
    dummy := head
    for dummy != nil {
        length++
        dummy = dummy.Next
    }

    return length
}
func rotateRight(head *ListNode, k int) *ListNode {
    length := lengthOfLL(head)

    if length == 0 || length == 1 {
        return head
    }

    k %= length
    if k == 0 {
        return head
    }

    slow, fast := head, head
    for k > 0 {
        fast = fast.Next
        k-=1
    }

    for fast.Next != nil {
        slow = slow.Next
        fast = fast.Next
    }

    newHead := slow.Next
    slow.Next = nil

    fast.Next = head
    return newHead
}